### ISSUE TITLE
Message duplication

### ISSUE THREAD

**Issue by chrisballinger** [ChatSecure/ChatSecure-iOS#906]
See intermittently when talking with a contact on ejabberd (with our server on prosody), on both incoming and outgoing plaintext messages.

**Comment by weiss** [comment]
I see duplication of incoming messages as well.  This is unrelated to the sender's server.  It happens when the recipient was offline while receiving messages.  ChatSecure goes online and then receives the messages both from offline storage and from MAM, and fails to deduplicate them.  (This bug hidden by enabling OMEMO or OTR, as the messages can only be decrypted once.)

The [deduplication logic][1] seems to classify a message as a duplicate [only if][2] `message.threadId == buddyUniqueId` is `true`.  Maybe this is the problem?  I don't understand the intention behind this check.

[1]: https://github.com/ChatSecure/ChatSecure-iOS/blob/01309217c1d84f97949a8718cf0ed0c291853629/ChatSecure/Classes/Controllers/XMPP/Storage/MessageStorage.swift#L138
[2]: https://github.com/ChatSecure/ChatSecure-iOS/blob/01309217c1d84f97949a8718cf0ed0c291853629/ChatSecure/Classes/Controllers/XMPP/Storage/MessageStorage.swift#L147

**Comment by ge0rg** [comment]
The issue description is utterly vague, so not sure if it is this issue or not. But I'm experiencing a similar problem to @weiss, with prosody as the server and ChatSecure as the only XMPP client used by the user. No encryption involved, most incoming messages are duplicated.

**Comment by tmolitor-stud-tu** [comment]
I get duplicates, too, every time ChatSecure reconnects to the xmpp server.

**Comment by Letterus** [comment]
Yep, I can confirm this issue. Offline messages (MUC) get duplicated each time I connect. Makes ChatSecure hard to use.

**Comment by theotix** [comment]
Same here

**Comment by ge0rg** [comment]
Tested with build 129 (4.3.1):

* sometimes messages still get duplicated when ChatSecure is offline (but not always?)
* now I get four "New Message" notifications per a single message, when the app was closed, or those four and in addition a fifth one with the actual body if the app is open.

**Comment by weiss** [comment]
Deduplication of offline vs. MAM messages still doesn't work for me with build 134 (4.3.4) either.

I still don't understand why messages are only classified as dups if [`foundMessage.threadId == buddyUniqueId`][1] is true.  Why isn't this a trivial `message.stanzaId` comparison for incoming and `message.originId` comparison for outgoing messages?  The [spec][2] guarantees uniqueness of stanza IDs, origin IDs are controlled by yourself, why do you need additional checks?

[1]: https://github.com/ChatSecure/ChatSecure-iOS/blob/11692533841276c17c0b029e65ab47cc30b6d476/ChatSecure/Classes/Controllers/XMPP/Storage/MessageStorage.swift#L147
[2]: https://xmpp.org/extensions/xep-0359.html

**Comment by chrisballinger** [comment]
I think originally it was written without oid/sid support, and it would discard duplicate elementIds sent by the same buddy, which turns out was a violation of the spec anyway. There is still a bug with how oid/sid are being stored in the database, which is why you'll see the query is searching elementId matching oid also. 

Without the check that the threadId matches, you could still have elementId collisions across message threads. It should probably at least be changed to `let threadId = message.threadId` though.

**Comment by ge0rg** [comment]
Issue still exists in 4.3.5 (135) from TestFlight.

**Comment by sgarringer** [comment]
Wondering if any progress is being made on this bug?  Prosody 0.10 and latest ChatSecure on iOS and every time I reconnect messages are duplicated.

**Comment by robled** [comment]
I've experienced this problem consistently since I switched from Android (Conversations client) to iOS (ChatSecure client).  The servers I connect to run ejabberd, and I found that when MUCs turn on the "archive" setting, then the dupes stop happening.  I hope this info helps since it seems to be related to settings specific to each MUC on a server.
